docs/guides/advanced-experimentation/anonymous-explainer.md

```diff
@@ -4,110 +4,90 @@ sidebar_position: 5
-        , LAG(ts) OVER (PARTITION BY user_id ORDER BY ts) as last_ts
-        , LEAD(ts) OVER (PARTITION BY user_id ORDER BY ts) as next_ts
-    FROM event_table
-)
+| Anon ID | User ID | Timestamp | Experiment | Variant
+| ------- | ------- | --------- | --------- | --------- | 
+anon1 | NULL | 2024-07-01 00:00:00 | new-checkout | control
+anon1 | 1 | 2024-07-01 10:00:00 | new-checkout |control
+anon1 | 2 | 2024-07-01 12:00:00 | new-checkout |control
+anon1 | NULL | 2024-07-02 12:00:00 | new-checkout | control
+anon2 | NULL | 2024-07-02 00:00:00 | new-checkout |treatment
+anon2 | 1 | 2024-07-02 10:00:00 | new-checkout |treatment
+anon3 | NULL | 2024-07-03 00:00:00 | new-checkout |control
+anon3 | 2 | 2024-07-03 10:00:00 | new-checkout |control
+anon4 | NULL | 2024-07-02 10:00:00 | new-checkout | treatment
 
-, anonymous_id_switch AS (
-    SELECT
-      *
-      , SUM(IF(last_anonymous_id != anonymous_id, 1, 0)) OVER (
-          PARTITION BY user_id
-          ORDER BY ts
-          BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW
-        ) AS cumulative_switch
-  FROM anonymous_id_lag
-)
+This example highlights how in general the relationship between User ID and Anonymous ID is many-to-many. That is, one Anonymous ID can be associated with multiple User IDs (a device is shared across two users) and one User ID might be associated with multiple Anonymous IDs (one user clears their cookies or has two devices).
 
-, user_login_windows_collapsed as (
-    SELECT
-      anonymous_id
-      , user_id
-      , cumulative_switch
-      , LOGICAL_OR(last_ts IS NULL) as is_first
-      , LOGICAL_OR(next_ts IS NULL) as is_last
-      , MIN(ts) as ts_min
-      , MAX(next_ts) as ts_max
-    FROM anonymous_id_switch
-    GROUP BY 1,2,3
-)
+Eppo can gracefully handle both this many-to-many relationship and assignment data with missing or duplicate records. When processing the assignment data, Eppo will perform the following clean up steps:
 
-SELECT
-    anonymous_id
-    , user_id
-    , IF(is_first, TIMESTAMP("0001-01-01 00:00:00"), ts_min) as ts_start_window
-    , IF(is_last, TIMESTAMP("9999-12-31 23:59:59"), ts_max) as ts_end_window
-FROM user_login_windows_collapsed
-ORDER BY user_id, ts_min;
+1. Filter to each Anonymous ID's first record during the experiment
+2. Check if the Anonymous ID was associated with more than one variant. If it was, remove it from downstream analysis
+3. For each Anonymous ID, determine the unique set of User IDs that was associated with it
+4. Join to both Anonymous ID and User ID metrics using the appropriate join column
+5. Aggregate metric events by Anonymous ID
 
-```
+Note that in step 3 we need to make sure each User ID is associated with only one Anonymous ID to avoid double counting post-authentication events (e.g., purchases). To ensure this, we associate each User ID with its *first* Anonymous ID during the course of the experiment.
```

Comment:
Should this be a numbered step above? Specifically, it sounds like there is a step before step 3 that is "For each User ID, identify the first record with a non-NULL Anonymous ID and exclude all other records."

I think we should clarify the ordering of the filtering too. For example, do we filter out duplicated rows for each User ID _before_  we filter out mixed anonymous IDs? I'm wondering if there's a case where this filtering step removes the record that would have made an Anonmous ID mixed.

Reply:
IMO it's more "how step 3 works" as opposed to a separate step

I don't think we dedupe records for User ID, just for anon ID. If user 1 has cookie A and cookie B, each with their own treatment, neither cookie A nor cookie B would be filtered out

Reply:
I think I'm confused then. This line makes it sound like we ARE deduplicating records for user id?

> Note that in step 3 we need to make sure each User ID is associated with only one Anonymous ID 

I'm struggling to reconcile that line with this scenario:

> If user 1 has cookie A and cookie B, each with their own treatment, neither cookie A nor cookie B would be filtered out

Doesn't this mean that a purchase from User 1 could be double-counted across cookie A or B? I feel like I'm missing something subtle.

Reply:
Oh yeah I think I'm getting tripped up. You're right, each cookie can only have multiple users, but each user can only have one cookie

So I see your point: do we check if a cookie is associated with multiple variants before we remove other userIDs or after

@jonathan-fulton probably can speak more to that part

Reply:
Hmm, I don't have a strong opinion, but it feels a little more intuitive to me to use a user's first non-mixed cookie. My understanding is that our approach follows from the assumption that these cookie-randomized experiments are generally independent across sessions. So, if a user has a non-mixed cookie that isn't impacted by a previous mixed cookie, it seems like we might as well include it.

Reply:
So after writing the SQL, it's a bit tricky to do the first non-mixed cookie.  Right now in the code we're just going to pick the first cookie a user is associated with.